### DISCUSSION TITLE
Get Refresh token from other origin

### DISCUSSION THREAD

**Discussion by NilsBaumgartner1994** [directus/directus#9347] [Archive]
I want to use directus on Domain A and having a Webapp on Domain B. 
The Webapp Domain B should offer the user to log in. The login should be able to use SSO (O_AUTH). But here is the point, Directus will only return the refresh_token as a HttpOnly Cookie, therefore i cant get the refresh_token on the Webapp.

How can i get the token ?

**Comment by rijkvanzanten** [comment]
> will only return the refresh_token as a HttpOnly Cookie, therefore i cant get the refresh_token on the Webapp.

Once you have the refresh token in the cookie of the current user session, you can use the /auth/refresh endpoint to request the actual access token 👍🏻

**Comment by NilsBaumgartner1994** [comment]
But that is the point. When I call the route from the webapp no cookies are sent :-(


Beste Grüße,
Nils Baumgartner

**Reply by rijkvanzanten** [reply]
How are you calling the route from the web app? Are you setting the correct flag to include cookies? Another thing to look into is the configuration for the cookie security itself, as browsers are getting a little tight nowadays.. You might have to enable Secure cookies (through REFRESH_TOKEN_COOKIE_SECURE) in order to allow the browser to send the cookie cross-origin.

**Reply by NilsBaumgartner1994** [reply]
When using the Webapp on the same origin as Directus, everything just works nice.
You can see my public directus here: https://se-services.informatik.uos.de/raiders/api/admin/login
Since it is a test server you can feel free to test/use it. The SSO "Studiptest" will accept you, whatever credentials you are using.

I use axios with:

axios.defaults.withCredentials = true
const api = axios.create({
	withCredentials: true
});

const response = await api.post('https://se-services.informatik.uos.de/raiders/api/auth/refresh', null, {
				"credentials": "include",
				"withCredentials": true
			});
			console.log(response);
			const accessToken = response.data.data.access_token;
			if(!!accessToken){
				setLoggedIn(true);
			}

You can see the project here:
https://github.com/FireboltCasters/Raiders-of-Storm/blob/main/Frontend/app/src/myComponents/auth/Login.tsx

**Reply by SeanDylanGoff** [reply]
Doesn't directly answer your question, but is the Webapp running on the same server? If yes you could also proxy directus to /api for example. This way you avoid all CORS issues entirely.

**Reply by NilsBaumgartner1994** [reply]
Nope they run on different servers.

for me to be clear:

This has to be set as following, or should i use "lax" ?
REFRESH_TOKEN_COOKIE_SAME_SITE:  "none"
REFRESH_TOKEN_COOKIE_SECURE: true

**Reply by NilsBaumgartner1994** [reply]
Sorry i meant different domains

**Reply by NilsBaumgartner1994** [reply]
The flow is currently:
- Users visits Webapp (Domain B):
- User selects one of the O_AUTH possibilities link with redirect to current Domain B
- Users enters credentials to O_AUTH: success
- Directus accepts it and redirects back to Domain B
- Webapp from Domain B wants Refresh_Token: fail
- No cookies were sent ...

**Comment by batuhanbilginn** [comment]
You should use the same server and set REFRESH_COOKIE_DOMAIN as root domain. 

Here is the example:
**Directus Backend:** api.example.com
**Frontend:** another.example.com
**Variable:** example.com

**Reply by NilsBaumgartner1994** [reply]
Wow a solution I haven't thought about... but that's not possible I have two different domains

**Reply by gkielwasser** [reply]
In addition to use ``useCredentials:true``, you have to take care Directus cookies envs are well set to allow cross domain cookies.

I have set those to have the cookie sent
``
REFRESH_TOKEN_COOKIE_SAME_SITE=none
REFRESH_TOKEN_COOKIE_SECURE=true
``

**Reply by NilsBaumgartner1994** [reply]
Done but it still dont want to

**Reply by gkielwasser** [reply]
Once Directus has redirected back to your WebApp, can you see in Chrome the cookie that is properly set ?

**Reply by NilsBaumgartner1994** [reply]
I see that a cookie is set.

WebApp Domain: 192.168.178.35


This cookie was set (where my Directus Backend runs):
Name: directus_refresh_token
Domain: se-services.informatik.uos.de
Path: /
Send for: Only Secure connection
Accessable for Script: No (only HTTP)

**Reply by NilsBaumgartner1994** [reply]
What am i missing here ?

**Reply by NilsBaumgartner1994** [reply]
Funny thing is: SameSite on the directus page is "None" but viewed from my WebApp it has the Value "Lax" ?
I am getting here different cookies wait

**Reply by NilsBaumgartner1994** [reply]
Yeah okay still the same:

I am getting two cookies:
One from my oauth2 name: oauth2.studip (SameSite: "Lax")
Directus exchanges this for the directus_refresh_token (SameSite "None")

But then every request i make it don't send cookies to directus from my webapp...

**Reply by NilsBaumgartner1994** [reply]
When I am on the same domain these cookies are sent and it works... but not on different domains.

**Reply by gkielwasser** [reply]
![refresh_token](https://user-images.githubusercontent.com/1532438/141438998-0f4a60f8-2351-41da-b4b7-f72e1cf6a70a.png)

Can you check again to validate your cookie, that from your webapp there is at least the directus_refresh_token set like in my screenshot.
In my case the domain (localhost) is the domain of my Directus instance. My webapp is also on another domain.

**Reply by NilsBaumgartner1994** [reply]
When visiting the directus instance.
<img width="1131" alt="Bildschirmfoto 2021-11-12 um 10 03 03" src="https://user-images.githubusercontent.com/33295329/141440384-1aef2d68-8602-4cd2-b4ff-6b5b91bda9c1.png">

On my webapp there is no cookie found in any request ... (running on localhost)

**Reply by NilsBaumgartner1994** [reply]
The refresh request from my webapp to directus...

<img width="669" alt="Bildschirmfoto 2021-11-12 um 10 07 11" src="https://user-images.githubusercontent.com/33295329/141441018-e011b592-f73e-43b6-a970-c203f4116e97.png">

**Reply by NilsBaumgartner1994** [reply]
> ![refresh_token](https://user-images.githubusercontent.com/1532438/141438998-0f4a60f8-2351-41da-b4b7-f72e1cf6a70a.png)
> 
> Can you check again to validate your cookie, that from your webapp there is at least the directus_refresh_token set like in my screenshot. In my case the domain (localhost) is the domain of my Directus instance. My webapp is also on another domain.

When opening my webapp (in Firefox) there are no cookies set (after redirect from directus)

<img width="1155" alt="Bildschirmfoto 2021-11-12 um 10 17 54" src="https://user-images.githubusercontent.com/33295329/141442468-b0dc49a4-d6cd-4796-804b-58d071f335ef.png">

**Reply by NilsBaumgartner1994** [reply]
The directus configuration:

directus:
    container_name: ${DOMAIN_PATH}directus
    env_file: .env
    image: directus/directus:latest
    expose:
      - ${BACKEND_PORT}
    volumes:
       # my volume configs
    networks:
      - directus
      - topDomainProxy
    depends_on:
      - cache
environment:
      KEY: 'XXXX'
      SECRET: 'XXXX'

      DB_CLIENT: 'sqlite3' # https://docs.directus.io/configuration/config-options/
      DB_FILENAME: './database/data.db'

      CACHE_ENABLED: ${BACKEND_CACHING}
      CACHE_STORE: 'redis'
      CACHE_REDIS: 'redis://cache:6379'

      REFRESH_TOKEN_COOKIE_SAME_SITE:  "none"
      REFRESH_TOKEN_COOKIE_SECURE: "true"
      CORS_ALLOWED_HEADERS: 'Authorization,Accept,Accept-Encoding,Accept-Language,Access-Control-Request-Headers,Access-Control-Request-Method,Cache-Control,Connec>

      # Make sure to set this in production
      # (see https://docs.directus.io/reference/environment-variables/#general)
      PUBLIC_URL: ${DOMAIN_PRE}://${HOST}/${DOMAIN_PATH}/${BACKEND_PATH}

      AUTH_PROVIDERS: "studip"
      .....

**Reply by NilsBaumgartner1994** [reply]
@batuhanbilginn and thanks for your help by the way :-)

**Comment by NilsBaumgartner1994** [comment]
When visiting the directus page there is a cookie correctly set.

<img width="1915" alt="Bildschirmfoto 2021-11-12 um 10 20 33" src="https://user-images.githubusercontent.com/33295329/141442867-6f29f843-37b9-45a6-af5a-4f675b0ddafe.png">

**Reply by gkielwasser** [reply]
About Directus envs, do you have those overriden or it is still the default values ?
```
CORS_METHODS
CORS_ALLOWED_HEADERS
```

Can you also show a screen showing all the request Headers sent with the refresh request ?

**Reply by NilsBaumgartner1994** [reply]
Yeah give me a moment

**Reply by NilsBaumgartner1994** [reply]
> CORS_METHODS
> CORS_ALLOWED_HEADERS

I had CORS_ALLOWED_HEADERS overriden with more allowings, but testet it by commenting it out.
So they also have the default values

> Can you also show a screen showing all the request Headers sent with the refresh request ?
I have also tried to copy the send fetch from the directus page.

<img width="1299" alt="Bildschirmfoto 2021-11-12 um 10 47 46" src="https://user-images.githubusercontent.com/33295329/141446733-74378361-0792-434d-b615-b1c0f91c21fc.png">

And this is my code for the refresh call

```
		try{
			const response = await api.post(getDirectusURL()+'/auth/refresh', undefined, {
				"credentials": "include",
				"withCredentials": true,
				"mode": "cors"
			});
			console.log(response);
			const accessToken = response.data.data.access_token;
			if(!!accessToken){
				setLoggedIn(true);
			}
		} catch (err){
			console.log(err);
		}
```

**Reply by NilsBaumgartner1994** [reply]
To compare, these are the headers on the directus page.

<img width="1438" alt="Bildschirmfoto 2021-11-12 um 10 49 31" src="https://user-images.githubusercontent.com/33295329/141447015-e4bcc51b-85a5-43fb-b9ab-8c11b1d44aed.png">

**Reply by gkielwasser** [reply]
Ok I have to say I don't see anything wrong so far..

I have also tested in Firefox just in case it would be a specific browser policy, but everthing works well.

Last thing, can you share the response headers you get from the refresh request?

**Reply by NilsBaumgartner1994** [reply]
<img width="578" alt="Bildschirmfoto 2021-11-12 um 12 56 39" src="https://user-images.githubusercontent.com/33295329/141463476-6df2ac91-f1b5-45ff-8544-bacfe9f5c823.png">

Message
`
{"errors":[{"message":"\"refresh_token\" is required in either the JSON payload or Cookie","extensions":{"code":"INVALID_PAYLOAD"}}]}
`

**Reply by NilsBaumgartner1994** [reply]
The Problem is, that no Cookies are beeing sent with the request...

**Reply by gkielwasser** [reply]
Yep your right my last question wasn't relevant.

The screenshot of the cookie is done from the Directus page itself. 

Can you confirm you also see the cookie displayed in Firefox debug from your app (192.168.178.35:19006) ?

**Reply by NilsBaumgartner1994** [reply]
> Can you confirm you also see the cookie displayed in Firefox debug from your app (192.168.178.35:19006) ?

On Firefox
I see the cookies are set on the directus page (see image below) in the refresh call
<img width="1552" alt="Bildschirmfoto 2021-11-12 um 13 41 01" src="https://user-images.githubusercontent.com/33295329/141468798-c5c673b2-3b6c-462b-b79d-edb737358a69.png">

But as soon i make the same call from my webapp (192.168.178.35:19006)
It is not sending the cookies

<img width="1599" alt="Bildschirmfoto 2021-11-12 um 13 41 49" src="https://user-images.githubusercontent.com/33295329/141468894-d26272c4-3275-43df-9bd4-efc9cccfa590.png">

**Reply by gkielwasser** [reply]
What I mean is on the webapp (your second screenshot)

Do you see the cookie on Firefox ? You can go on: Storage -> Cookies

**Reply by NilsBaumgartner1994** [reply]
Yep, is empty ...

**Reply by NilsBaumgartner1994** [reply]
But I think it should be possible to send Cookies from Domain A (http) to Domain B (https)?

**Reply by gkielwasser** [reply]
From what I see, the issue is not related to the refresh call or to the fact the cookie is not attached to the request.

To simplify the issue you could just ensure you see the cookie set on the Directus page (it is the case if I understood), then go on a simple page of your web app that is not doing any operations. You should see already the cookie in Firefox storage.

I would use a only html page to validate this step.

Also you should take care Firefox is not displaying any warning about cookie. When it drops some cookie, for any security reason, it usually display a warning message.

**Reply by NilsBaumgartner1994** [reply]
I created a small html page:

```
<html>
<body>
<h1>Directus Cookie test</h1>
</body>
</html
```

Served it with "serve" (this time on port 5000)
And no Cookie found on the webapp

<img width="1226" alt="Bildschirmfoto 2021-11-12 um 14 44 59" src="https://user-images.githubusercontent.com/33295329/141476782-008eda79-e3fc-4f91-a574-e1d1f7c30ec2.png">

But of course i am logged in at directus admin page

<img width="1270" alt="Bildschirmfoto 2021-11-12 um 14 46 31" src="https://user-images.githubusercontent.com/33295329/141477031-90472910-82be-4fb3-a686-d629b5121dfa.png">

I am asking again:
Is it correct, that I can use a non https page (http) to see the cookies of an https page ?

**Reply by NilsBaumgartner1994** [reply]
Because i also don't see the directus cookie at any other page like for example: https://www.oneplus.com

I am calling at directus: https://se-services.informatik.uos.de/myapp/api/auth/login/studip?redirect=http://192.168.178.35:5000/
So that it should redirect me to my webapp

**Reply by gkielwasser** [reply]
Did you validate the cross cookie is properly set when everything is deployed and using only https ?

**Reply by NilsBaumgartner1994** [reply]
I am sorry, can you explain what exactly you mean ?

My Webapp offers a link which goes to: https://se-services.informatik.uos.de/myapp/api/auth/login/studip?redirect=http://192.168.178.35:19006/myapp/app/login

So, that directus will redirect me back to my app.

After that the following cookies were set 

(when visiting directus):
<img width="1131" alt="Bildschirmfoto 2021-11-12 um 15 18 25" src="https://user-images.githubusercontent.com/33295329/141481384-d64f2bb2-435c-4f6e-8982-0e12a0bd067a.png">

(when visiting my page):
None

My Webapp is running on http and directus is running on https on different domains.

**Reply by gkielwasser** [reply]
I'm not sure of your final goal, is it for development purpose to redirect from https to http ?

In case it is possible, I would fully develop the cross domain authentication using http and then test with both domains in https.

**Reply by NilsBaumgartner1994** [reply]
My goal is, to use directus on an other domain than the webapp.
Therefore I need to log in via the webapp (for example via oAuth2) from the webapp and being in the end redirected to it back.

**Reply by gkielwasser** [reply]
> My goal is, to use directus on an other domain than the webapp. Therefore I need to log in via the webapp (for example via oAuth2) from the webapp and being in the end redirected to it back.

I think we have fully covered this :). I mean using cross origins, and it should work in case you work http -> http and https -> https.

I would say you cannot currently solve with cookies your case due to http/https.
 
You should follow [this](https://github.com/directus/directus/discussions/8867). 
I think you are enforced to use Bearer authentication in your case.

OAuth flow would redirect to http://192.168.178.35:5000?refresh_token=

**Reply by NilsBaumgartner1994** [reply]
That would be totatly fine to use whatever but just to have be able to use it on different domains.
Then how would be the call to use a provided oAuth2 flow of directus ?

**Reply by gkielwasser** [reply]
I'm not sure of your question, do you mean if [this ticket](https://github.com/directus/directus/discussions/8867) would be done ?

You would just do as your are currently doing, except you would call the refresh endpoint and give as a parameter the "refresh_token" retrieved from the redirection.

**Reply by NilsBaumgartner1994** [reply]
So that would be like the following:

Webapp --> directus /oauth/provider?redirect=/customEndpoint --> oAuth Provider --> /customEndpoint?redirect=webapp --> webapp?refresh_token=token 

So with the custom endpoint we would "hack" us to give the webapp the refresh_token ?

**Reply by gkielwasser** [reply]
Login:
1 - Webapp -> directus/login/oauth/:provider?redirect=http://192.168.178.35:5000?refresh_token
2- [...] all auth flow is managed on directus (including 2fa if there is)
3- at the end redirection is done to your webapp
4- webapp can read refresh_token from query parameters

Note: There could be also a variant using a popup. I'm using it with cookies currently.

All steps are done in a popup. The webapp itself does not redirect. At the end the popup close and send the refresh_token to the webapp.

**Reply by NilsBaumgartner1994** [reply]
Can you post your example with cookies ?

**Reply by NilsBaumgartner1994** [reply]
> Login: 1 - Webapp -> directus/login/oauth/:provider?redirect=http://192.168.178.35:5000?refresh_token 2- [...] all auth flow is managed on directus (including 2fa if there is) 3- at the end redirection is done to your webapp 4- webapp can read refresh_token from query parameters
> 
> Note: There could be also a variant using a popup. I'm using it with cookies currently.
> 
> All steps are done in a popup. The webapp itself does not redirect. At the end the popup close and send the refresh_token to the webapp.

I think that this does not work...

My public testserver:
https://se-services.informatik.uos.de/myapp/api/admin/login

Now visiting, as you provided under: 1: Webapp -> directus/login/oauth/:provider?redirect....
https://se-services.informatik.uos.de/myapp/api/auth/login/studiptest?redirect=http://localhost?refresh_token

Filling any credentials you like (accepting any provided), you will land on localhost without the refresh_token.

**Comment by NilsBaumgartner1994** [comment]
https://gist.github.com/NilsBaumgartner1994/6e1a764491c736d7f187017c9e1f17cc
